network/src/protocols/discovery/mod.rs

```diff
@@ -382,217 +383,208 @@ where
+pub struct Note {
+    /// Id of the peer.
     peer_id: PeerId,
+    /// The validator node's signed `PeerInfo`.
+    signed_peer_info: SignedPeerInfo,
+    /// The validator node's signed `FullNodePayload`.
+    signed_full_node_info: SignedFullNodeInfo,
+}
+
+impl Note {
+    fn new(
+        signer: &Signer,
+        peer_id: PeerId,
+        addrs: Vec<Multiaddr>,
+        dns_seed_addr: &[u8],
+        epoch: u64,
+    ) -> Self {
+        let peer_info = PeerInfo { addrs, epoch };
+        let signed_peer_info = peer_info.sign(&signer);
+
+        let full_node_info = FullNodeInfo {
+            dns_seed_addr: dns_seed_addr.to_vec(),
+            epoch,
+        };
+        let signed_full_node_info = full_node_info.sign(&signer);
+
+        Note {
+            peer_id,
+            signed_peer_info,
+            signed_full_node_info,
+        }
+    }
+
+    /// Verifies validity of notes. Most fields have already been verified during initial
+    /// deserialization, so this only verifies that the note contains properly signed infos.
+    fn verify(&self, pub_key: &Ed25519PublicKey) -> Result<(), NetworkError> {
+        self.signed_peer_info.verify(&pub_key)?;
+        self.signed_full_node_info.verify(&pub_key)?;
+        Ok(())
+    }
+
+    /// Shortcut to the addrs embedded within the Note
+    fn addrs(&self) -> &Vec<Multiaddr> {
+        &self.signed_peer_info.peer_info.addrs
+    }
+
+    /// The current implementation derives epoch from the PeerInfo.
+    fn epoch(&self) -> u64 {
+        self.signed_peer_info.peer_info.epoch
+    }
+}
+
+/// A `PeerInfo` authenticated by the peer's root `network_signing_key` stored on-chain.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct SignedPeerInfo {
+    /// A `PeerInfo` represents the network address(es) of a Peer.
+    peer_info: PeerInfo,
+    /// A signature over the above serialzed `PeerInfo`, signed by the validator's
+    /// `network_signing_key` referred to by the `peer_id` account address.
+    signature: Ed25519Signature,
+}
+
+impl SignedPeerInfo {
+    fn verify(&self, pub_key: &Ed25519PublicKey) -> Result<(), NetworkError> {
+        let peer_info_bytes =
+            lcs::to_bytes(&self.peer_info).map_err(|_| NetworkErrorKind::ParsingError)?;
+        self.signature
+            .verify(&get_hash(&peer_info_bytes), &pub_key)
+            .map_err(|_| NetworkErrorKind::SignatureError)?;
+        Ok(())
+    }
+}
+
+/// A `PeerInfo` represents the network address(es) of a Peer.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct PeerInfo {
+    /// Network addresses this peer can be reached at.
     addrs: Vec<Multiaddr>,
+    /// Monotonically increasing incarnation number used to allow peers to issue
+    /// updates to their `PeerInfo` and prevent attackers from propagating old
+    /// `PeerInfo`s. This is usually a timestamp.
     epoch: u64,
-    /// the raw `Note` sent from remote
-    raw_note: Note,
```

Comment:
:fire:

Reply:
My preference would be that we remove all expect from this code base. Could we punt on this for now?

Reply:
@davidiw expect/unwrap are not inherently bad and a blanket statement like "we shouldn't allow them in the codebase" isn't helpful IMO. Whether to use them or not depends on the situation and the invariants that need to be upheld.

Reply:
Many of the expect statements assume symmetry, because I can safely deserialize it, I can safely serialize it. Or that my signing key will always sign. We have to really understand our interfaces to be confident that they are infallible or that the system should halt. I can imagine an attack vector where someone can craft a message that allows deserialization but might not be serializable. At the same time, well placed asserts can halt execution where safety or system properties have been violated. Given that this is part of the active run time, I feel a bit more conflicted about using panics.

Fwiw, I was referring to this explicit file not the general Libra code base, though admit the language choice on my behalf was a bit ambiguous.